File: Treachery-iOS/Treachery-iOSTests/GameModeTests.swift
import Testing
import Foundation
@testable import Treachery_iOS

struct GameModeTests {

    @Test func allCases() {
        #expect(GameMode.allCases.count == 4)
    }

    @Test func rawValues() {
        #expect(GameMode.treachery.rawValue == "treachery")
        #expect(GameMode.planechase.rawValue == "planechase")
        #expect(GameMode.treacheryPlanechase.rawValue == "treachery_planechase")
        #expect(GameMode.none.rawValue == "none")
    }

    @Test func displayNames() {
        #expect(GameMode.treachery.displayName == "Treachery")
        #expect(GameMode.planechase.displayName == "Planechase")
        #expect(GameMode.treacheryPlanechase.displayName == "Both")
        #expect(GameMode.none.displayName == "Life Tracker")
    }

    @Test func includesTreachery() {
        #expect(GameMode.treachery.includesTreachery == true)
        #expect(GameMode.treacheryPlanechase.includesTreachery == true)
        #expect(GameMode.planechase.includesTreachery == false)
        #expect(GameMode.none.includesTreachery == false)
    }

    @Test func includesPlanechase() {
        #expect(GameMode.planechase.includesPlanechase == true)
        #expect(GameMode.treacheryPlanechase.includesPlanechase == true)
        #expect(GameMode.treachery.includesPlanechase == false)
        #expect(GameMode.none.includesPlanechase == false)
    }

    @Test func decodesFromJSON() throws {
        let json = Data(#""treachery_planechase""#.utf8)
        let mode = try JSONDecoder().decode(GameMode.self, from: json)
        #expect(mode == .treacheryPlanechase)
    }
}

File: TreacheryAndroid/app/src/test/java/com/solomon/treachery/model/GameModeTests.kt
package com.solomon.treachery.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class GameModeTests {

    @Test
    fun `all cases exist`() {
        assertEquals(4, GameMode.entries.size)
    }

    @Test
    fun `raw values match expected strings`() {
        assertEquals("treachery", GameMode.TREACHERY.value)
        assertEquals("planechase", GameMode.PLANECHASE.value)
        assertEquals("treachery_planechase", GameMode.TREACHERY_PLANECHASE.value)
        assertEquals("none", GameMode.NONE.value)
    }

    @Test
    fun `display names are human readable`() {
        assertEquals("Treachery", GameMode.TREACHERY.displayName)
        assertEquals("Planechase", GameMode.PLANECHASE.displayName)
        assertEquals("Both", GameMode.TREACHERY_PLANECHASE.displayName)
        assertEquals("Life Tracker", GameMode.NONE.displayName)
    }

    @Test
    fun `includesTreachery is correct`() {
        assertTrue(GameMode.TREACHERY.includesTreachery)
        assertFalse(GameMode.PLANECHASE.includesTreachery)
        assertTrue(GameMode.TREACHERY_PLANECHASE.includesTreachery)
        assertFalse(GameMode.NONE.includesTreachery)
    }

    @Test
    fun `includesPlanechase is correct`() {
        assertFalse(GameMode.TREACHERY.includesPlanechase)
        assertTrue(GameMode.PLANECHASE.includesPlanechase)
        assertTrue(GameMode.TREACHERY_PLANECHASE.includesPlanechase)
        assertFalse(GameMode.NONE.includesPlanechase)
    }

    @Test
    fun `fromValue returns correct mode`() {
        assertEquals(GameMode.TREACHERY, GameMode.fromValue("treachery"))
        assertEquals(GameMode.PLANECHASE, GameMode.fromValue("planechase"))
        assertEquals(GameMode.TREACHERY_PLANECHASE, GameMode.fromValue("treachery_planechase"))
        assertEquals(GameMode.NONE, GameMode.fromValue("none"))
    }

    @Test
    fun `fromValue returns null for unknown value`() {
        assertNull(GameMode.fromValue("unknown"))
        assertNull(GameMode.fromValue(""))
    }
}
